content/shmem_malloc_hints.tex

```diff
@@ -0,0 +1,97 @@
+
+\apisummary{
+    Collective memory allocation routine with support for providing hints.
+}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+void *@\FuncDecl{shmem\_malloc\_with\_hints}@(size_t size, long hints);
+\end{Csynopsis}
+
+\begin{apiarguments}
+    \apiargument{IN}{size}{The size, in bytes, of a block to be
+        allocated from the symmetric heap. This argument is of type \CTYPE{size\_t}}
+    \apiargument{IN}{hints}{A bit array of hints provided by the user to the implementation}
+\end{apiarguments}
+
+
+\apidescription{
+
+    The \FUNC{shmem\_malloc\_with\_hint} like \FUNC{shmem\_malloc} routine returns a pointer to a block of at least
+    \VAR{size} bytes, which shall be suitably aligned so that it may be
+    assigned to a pointer to any type of object.  This space is allocated from
+    the symmetric heap (similar to \FUNC{shmem\_malloc}).  When \VAR{size} is zero, 
+    the \FUNC{shmem\_malloc\_with\_hint} routine performs no action and returns a null pointer. 
+    
+    In addition to the \VAR{size} argument, the \VAR{hint} argument is provided by the user. 
+    The \VAR{hint} describes the expected manner in which the \openshmem program may use the allocated memory.
+    The valid usage hints are described in Table~\ref{usagehints}. Multiple hints are expressed as \CONST{OR} of \VAR{hints}.
+    
+    The information provided by the \VAR{hint} is used to optimize for performance by the implementation. 
+    If the implementation cannot optimize, the behavior is same as \FUNC{shmem\_malloc}.
+    If more than one hint is provided, the implementation will make the best effort to use one or more hints 
+    to optimize performance. 
+            
```

Comment:
Must all PEs allocate from the same special memory?  It's not clear if asymmetry can exist. Does this impose additional implicit synchronization for each subset configuration of hints if it cannot satisfy the entire hint list?

Also, what happens if you OR `SHMEM_HINT_NONE` with other hint behaviors?

Reply:
Good question. 

All use cases that I have thought of requires the memory to be symmetric (same kind of memory). 

Regarding extra synchronization, it depends on the implementation. If the implementations maintain asymmetric memory sizes (say each PE starts with different amount of special memory) on the PEs, you might need the extra synchronization for agreement. Otherwise, I do not see a need. In a way, it is similar to current DRAM allocations. Also, for the implementation we explored, we did not need extra synchronization.

I’m reluctant to add such a constraint. Without such constraint, the implementations are free to explore either approach (symmetric and asymmetric). 

Do see value in specifying one way or other?

Reply:
To clarify, what would using `long hint = SHMEM_HINT_NONE | SHMEM_HINT_LOW_LAT_MEM | SHMEM_HINT_HIGH_BW_MEM` do?

Dropping `SHMEM_HINT_NONE`, what if the platform could provide `SHMEM_HINT_LOW_LAT_MEM` or `SHMEM_HINT_HIGH_BW_MEM` but not both simultaneously?  Will we see application code marked up like this because who doesn't want to use low latency and high bandwidth memory for their application? Does the "best effort" default to a platform-specific precedence? The only feedback is that an allocation succeeded or it did not.

Reply:
> To clarify, what would using long hint = SHMEM_HINT_NONE | SHMEM_HINT_LOW_LAT_MEM | SHMEM_HINT_HIGH_BW_MEM do?

Though this is a legal usage, it does not make sense to use. The implementations are allowed to default to shmem_malloc in this case. 

> Dropping SHMEM_HINT_NONE, what if the platform could provide SHMEM_HINT_LOW_LAT_MEM or SHMEM_HINT_HIGH_BW_MEM but not both simultaneously? Will we see application code marked up like this because who doesn't want to use low latency and high bandwidth memory for their application? Does the "best effort" default to a platform-specific precedence? The only feedback is that an allocation succeeded or it did not.

```
“If more than one hint is provided, the implementation will make the best effort to use one or more hints to optimize performance. 
“
```
My intention with this statement was to provide flexibility for the implementations to optimize as they wish when the user provides multiple hints. Obviously, some combinations of hints might not make sense. In such cases, If the implementations want to give precedence of one hint over others, the proposal allows it. That (assigning priorities to hints) is one way to implement it, but not the only way.